File: Exercises/Lecture1/VanDerPol.m
%% Driver

mu = 10;
x0 = [2.0; 0.0];
options = odeset('Jacobian',@JacVanDerPol,'RelTol',1.0e-6,'AbsTol',1.0e-6);
[T,X]=ode15s(@VanDerPolModel,[0 5*mu],x0,options,mu);

%% Model

function xdot = VanDerPolModel(t,x,mu)
% VANDERPOL Implementation of the Van der Pol model
%
% Syntax: xdot = VanDerPol(t,x,mu)
xdot=zeros(2,1);
xdot(1) = x(2);
xdot(2) = mu*(1-x(1)*x(1))*x(2)-x(1);
end

function Jac = JacVanDerPol(t,x,mu)
% JACVANDERPOL Jacobian for the Van der Pol Equation
%
% Syntax: Jac = JacVanDerPol(t,x,mu)

Jac = zeros(2,2);
Jac(2,1) = -2*mu*x(1)*x(2)-1.0;
Jac(1,2) = 1.0;
Jac(2,2) = mu*(1-x(1)*x(1));
end

File: Exercises/Lecture1/VanDerPol.jl
using DifferentialEquations
using Plots

## Initialize variables

μ = 10.0
x₀ = [2.0 0.0]
tspan = (0.0, 5.0*μ)

## Model

function VanDerPol(ẋ,x₀,tspan,μ)
# VANDERPOL Implementation of the Van der Pol model
#
# Syntax: ẋ = VanDerPol(x,t,μ)
    ẋ = zeros(2)
    x₁, x₂ = x₀
    ẋ[1] = x₂
    ẋ[2] = μ*(1-x₁*x₁)*x₂-x₁
    return ẋ
end

function JacVanDerPol(Jac,x₀,tspan,μ)
# JACVANDERPOL Jacobian for the Van der Pol Equation
#
# Syntax: Jac = JacVanDerPol(x,t,μ)
    Jac = zeros(2, 2)
    x₁, x₂ = x₀
    Jac[2,1] = -2*μ*x₁*x₂-1.0
    Jac[1,2] = 1.0
    Jac[2,2] = μ*(1-x₁*x₁)
    return Jac
end

ẋ = ODEFunction(VanDerPol;jac=JacVanDerPol)

## Define the problem and solve

prob = ODEProblem(ẋ,x₀,tspan,μ)
sol = solve(prob, QNDF(), reltol=1e-6, abstol=1e-6)

## Data Visualization

plot(sol)
